README.md
# aiida-sssp-workflow

The `aiida-sssp-workflow` is an aiida plugin to run the verification for a given pseudopotential. The plugin contains
workflows to verify the pseudopotential.
It can:

1) evaluate the [delta factor]() of the pseudopotential with respect to WIEN2K all-electrons results.
2) Converge test on varies of properties to give a recommended energy cutoff of the pseudopotential, include properties:
    1) Cohesive energy
    2) phonon frequencies
    3) pressure
    4) bands distance

### The computational details to running the calculation

#### Input Structures:

- In Δ-factor calculation: most stable elemental system from [Cottenier's work](http://molmod.ugent.be/deltacodesdft)
    and rare-earth nitrides from [Topsakal-Wentzkovitch work](https://www.sciencedirect.com/science/article/abs/pii/S0927025614005059);
- Phonon, pressure, cohesive energy: Cottenier's structures
    (except SiF4 has been used for F because of convergence issues) and
    rare-earth nitrides; Use primitive cells.
- Bands: Cottenier's structures reduced to primitive cells
    (except SiF4 has been used for F because of convergence issues) and rare-earth nitrides.
    PwbandWorkflow will make a primitive cell for band calculation (Remember to turn off the relax step).

#### Parameters of Δ calculations

- wave function cutoffs: 200 Ry;
- dual = 8 (PAW/US), 4 (NC); Mn/Fe/Co have larger duals tested as well; 12 and 16.
    We have gone in a mode where we do not use the dual, but we use ECUTRHO and ECUTWFC. However, dual is still used in
    simply setting the ecutwfc/ecutrho pairs.
- k-points: 0.1A^-1;
- smearing (degauss): Marzari-Vanderbilt, 0.01 Ry;
- non spin-polarized calculations except Mn (antiferromagnetic),
    O and Cr (antiferromagnetic),
    Fe, Co, and Ni (ferromagnetic).

> As for calculation of lanthenide, always increase `nbnd` to two times of the default number.

#### Parameters in phonon, pressure, cohesive energy calculations:

- k-points: 0.15A^-1
- smearing: Marzari-Vanderbilt, 0.01 Ry;
- k-points for the isolated atoms: 1x1x1;
- smearing for the isolated atoms: gaussian 0.01 Ry;
- unit cell for the isolated atoms: 12x12x12 Å with atom sit in [6.0, 6.0, 6.0] the middle of the cell;
- q-point: only calculate the phonon frequencies on Brillouin-Zone border q=(0.5, 0.5, 0.5).
- all calculations non-spin-polarized.

> In isolate atom energy calculation of cohesive energy evaluation.
> As for lanthenide, increase `nbnd` to three times of the default number. Moreover, use more RAM(by increase `num_machine` to 4).


> NOTE: PWscf writes in the output something called total energy. This is *NOT* the total energy when you have smearing;
> it’s the total free energy E-TS. PWscf also writes -TS, so one can get back the total energy E.
> In general (for a metal) E-TS should be used. For an atom instead the total energy should be used,
> since the -TS term is not really physical (it comes from the entropy of fractional occupations on the atom).
> Check with Nicola if you have atoms where -TS is different from zero. (http://theossrv1.epfl.ch/Main/ElectronicTemperature)


##### The convergence pattern for the phonons is calculated as:
- circle = (1/N * ∑i=1,N [ωi(cutoff) - ωi(200Ry)]2 / ωi(200Ry)2)1/2 * 100 (in percentage) and half error bar = Max |[ω(cutoff) - ω(200Ry)] / ω(200Ry)| * 100, if the highest frequency is more than 100 cm-1;
- circle = (1/N * ∑i=1,N [ωi(cutoff) - ωi(200Ry)]2)1/2 (absolute value) and half error bar = Max |ωi(cutoff) - ω(200Ry)|, if the highest frequency is less than 100 cm-1;
- N is the total number of frequencies;
- For some elements, we have neglected the first n frequencies in the summation above, because the frequencies are negative and/or with strong oscillations as function of the cutoff for all the considered pseudos). We have neglected the first four frequencies for H and I, 12 for N and Cl, 6 for O and SiF4 (which replaces F).

#### Bands calculations:

- k-points for the self-consistent calculation: 0.1; (can use cache one for the latter calculation)
- k-points for the bands calculation (as in, calculations of the eta and eta10 factors): uniform mesh 0.2 with no symmetry reduction, rather than high-symmetry path which is not determinant;
- smearing: Marzari-Vanderbilt, 0.01 Ry in scf calculation and Fermi-Dirac in bands distance calculation;
- all calculations non spin-polarized.

## Repository contents

## Features

## More meta-info collection

### SiF4 structure and its (V0, B0, B1) reference value
Re-generate the SiF4 structure start from the cif file from [COD database](http://www.crystallography.net/cod/index.php). Detail inputs parameters are list below.

#### Pseudopotentials(SSSP-v1.1 precision)
- Si: Si.pbe-n-rrkjus_psl.1.0.0.upf
- F: F.oncvpsp.upf

#### Pw relax and eos

##### pwscf parameters
```
'SYSTEM': {
    'degauss': 0.0045,
    'ecutrho': 1600,
    'ecutwfc': 200,
    'occupations': 'smearing',
    'smearing': 'cold',
},
'ELECTRONS': {
    'conv_thr': 1e-10,
},
```

##### EOS parameters

- seven points
- 0.02 interval

## For maintainers

To create a new release, clone the repository, install development dependencies with `pip install '.[dev]'`, and then execute `bumpver update`.
This will:

  1. Create a tagged release with bumped version and push it to the repository.
  2. Trigger a GitHub actions workflow that creates a GitHub release.

Additional notes:

  - Use the `--dry` option to preview the release change.
  - The release tag (e.g. a/b/rc) is determined from the last release.
    Use the `--tag` option to switch the release tag.

## License

MIT


## Contact

📧 email: jusong.yu@epfl.ch


CHANGELOG.md
## aiida-sssp-workflow

### 0.2.0
- Add rho and wfc convergence test workflows in convergence verification at the same time
- Delta factor verification on oxides and compare the results with AE
### 0.2.0-beta.1
Fix some issues with running verification on eiger and kindly as the fixes for the production release.

### 0.2.0-beta.0
the release are ready to be run by command line interface.
Will be the first stable backend of aiidalab-sssp app.

- The legacy convergence workchains are all extended from _base convergence work chain.
- The pseudo_parser library is add for pseudopotential parsing only.
- Add CLI interface for verification workflow and the properties can be chosen.
- Test verification (at `test` protocol) for silicon, fluorine(SiF4 in convergence WF), gold and a lanthenide element Samarium (Sm). Fix some issues.

### 0.2.0-alpha.1
The release to correctly deploy the documentation and install the package in aiidalab

- replace fortran efermi with pure python efermi [#50](https://github.com/aiidateam/aiida-sssp-workflow/pull/50)
- verification workflow based on legacy workflows
- delta factor workflow accept dual as input

### 0.2.0-alpha.0
The release after Jason start working at EPFL

Add Legacy convergence workflows and simplify the interface of all workflows

- use aiida-pseudo for `UpfData`.
- add legacy convergence workflows for comparing the old sssp and for benchmarking purpose
- add `BandsDistanceWorkChain` for future bands chessboard plot

### 0.1.0
The first release

fixbug for the workflows and restructure the repository

- add bandstructure evaluation to verification
- mag_structure generate bugfix for Fe,Co,Ni

### 0.1.0b0
The initial (pre) release contain the most workflows

- workflow delta_measure
- the cli tool for delta factor calculation
- the convergence workflow integrate with aiida-optimize
- workflow convergence cohesive energy
- workflow convergence phonon frequencies
- workflow convergence pressue
- workflow convergence bands distance


examples/example_eos.py
#!/usr/bin/env python
"""
Running EquationOfStateWorkChain example
You can import the necessary nodes of examples-node-archive
(and config pw-6.6 code) to run this example.
"""
from aiida.common import AttributeDict
from aiida import orm
from aiida.engine import run_get_node

from aiida.plugins import WorkflowFactory

EOSWorkChain = WorkflowFactory('sssp_workflow.eos')


def run_eos(code, structure, upf):
    """eos run for silicon"""
    builder = EOSWorkChain.get_builder()
    inputs = AttributeDict({
        'structure': structure,
        'scale_count': orm.Int(7),
        'scf': {
            'pw': {
                'code':
                code,
                'pseudos': {
                    upf.element: upf
                },
                'parameters':
                orm.Dict(
                    dict={
                        'SYSTEM': {
                            'degauss': 0.02,
                            'ecutrho': 800,
                            'ecutwfc': 200,
                            'occupations': 'smearing',
                            'smearing': 'marzari-vanderbilt',
                        },
                        'ELECTRONS': {
                            'conv_thr': 1e-10,
                        },
                    }),
                'metadata': {
                    'options': {
                        'resources': {
                            'num_machines': 1
                        },
                        'max_wallclock_seconds': 1800,
                        'withmpi': True,
                    },
                },
            },
            'kpoints_distance': orm.Float(0.1),
        }
    })
    builder.update(**inputs)
    res, node = run_get_node(builder)

    return res, node


if __name__ == '__main__':
    from aiida.orm import load_code, load_node

    code = load_code('qe-6.6-pw@daint-mc')

    # Silicon eos
    structure = load_node('0c5793b9-e52c-4708-b5a1-f53b5c4814bf'
                          )  # structure of silicon in ground state
    upf = load_node(
        '9d9d57fc-49e3-4e0c-8c37-4682ccc0fb51')  # upf node of silicon

    res, node = run_eos(code, structure, upf)
    print(res)  # node.outputs
    print(node)

    # Print eos result
    print(node.outputs.output_parameters.get_dict())
